service/lxd.go

```diff
@@ -756,18 +735,11 @@ func (s *LXDService) waitReady(ctx context.Context, c lxd.InstanceServer, timeou
 		}
 	}()
 
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
-	defer cancel()
-
-	if timeoutSeconds > 0 {
-		select {
-		case <-finger:
-			break
-		case <-ctx.Done():
-			return fmt.Errorf("LXD is still not running after %ds timeout (%v)", timeoutSeconds, errLast)
-		}
-	} else {
-		<-finger
+	select {
+	case <-finger:
+		break
+	case <-ctx.Done():
+		return fmt.Errorf("Timed out waiting for LXD to start: %w", errLast)
```

Comment:
There is one error in the pipeline showing ` Error: Timed out waiting for LXD to start: %!w(<nil>)`. Should we instead show the context's error using `context.Cause(ctx)`? 

Maybe within `waitReady` we can derive another context which can be cancelled by the goroutine using a custom error (lastError if present). You can use `WithCancelCause` when deriving the new context.

Reply:
OK I've made it a bit more complicated but it eases up the UX a bit:

The `finger` channel is gone now. Instead there's 2 contexts:
The main context has a 1 minute deadline when the function is invoked, similar to before. There is now an additional `unixCtx` that waits up to 5 seconds for the LXD unix socket to appear. If this context's deadline is exceeded and we still haven't found the unix socket, the error returned from the API is used to cancel the main context.

The function blocks on the main context, and will return an error if the context cause is either `DeadlineExceeded`, or _not_ `Canceled`, which implies the issue was detecting the unix socket.

Reply:
Ah, wanted to follow up on this thread, see https://github.com/canonical/microcloud/pull/455#discussion_r1834602451.